### PR TITLE
Feat/peer events

### PR DESCRIPTION
**Pull request by siphiuel**
Listen to `newPeer`/`peerDead` events originating from PubSub. Includes a small AddPeer/SelectPeer refactor.
Add `IsOnline`, `HasHistory`, and `GetPeerStats` methods to WakuNode.
Add a `connStatusChan` parameter to WakuNode that will receive `ConnStatus(IsOnline, HasHistory)` structs.
Add peer event handling to Filter protocol (and maybe more).

### CODE REVIEW

**Review by richard-ramos — COMMENTED** [review]
Great work!

I left some minor comments

File: examples/filter2/go.mod [hunk 1]
```diff
@@ -10,3 +10,5 @@ require (
 )
 
 replace github.com/status-im/go-waku => ../..
+
+replace github.com/status-im/go-wakurelay-pubsub => ../../../go-wakurelay-pubsub
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply examples/filter2/go.mod:14]: Done.

File: examples/peer_events/go.mod [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+require (
+	github.com/ethereum/go-ethereum v1.9.5
+	github.com/ipfs/go-log v1.0.5
+	github.com/multiformats/go-multiaddr v0.3.1 // indirect
+	github.com/status-im/go-waku v0.0.0-20210525134543-f6ebb5b0f9ae
+)
+
+replace github.com/status-im/go-waku => ../..
+
+replace github.com/status-im/go-wakurelay-pubsub => ../../../go-wakurelay-pubsub
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply examples/peer_events/go.mod:14]: Removed replace altogether, as PR is merged now.

File: waku/v2/utils/peer.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	// This will require us to check for various factors such as:
+	//  - which topics they track
+	//  - latency?
+	//  - default store peer?
+
+	// Selects the best peer for a given protocol
+
+	var peers peer.IDSlice
+	for _, peer := range host.Peerstore().Peers() {
+		protocols, err := host.Peerstore().SupportsProtocols(peer, protocolId)
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply waku/v2/utils/peer.go:23]: It's the same thing: https://github.com/libp2p/go-libp2p/blob/8ec1eec8f83e472ffc13a386780d8a9ff75edf4a/p2p/host/basic/basic_host.go#L543

File: examples/peer_events/go.mod [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+require (
+	github.com/ethereum/go-ethereum v1.9.5
+	github.com/ipfs/go-log v1.0.5
+	github.com/multiformats/go-multiaddr v0.3.1 // indirect
+	github.com/status-im/go-waku v0.0.0-20210525134543-f6ebb5b0f9ae
+)
+
+replace github.com/status-im/go-waku => ../..
+
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply examples/peer_events/go.mod:13]: Done

File: waku/v2/utils/peer.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		if err != nil {
+			log.Error("error obtaining the protocols supported by peers", err)
+			return nil
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply waku/v2/utils/peer.go:26]: Haha, you're too gracious. That's a dumb mistake! ) Fixed.

File: waku/v2/protocol/relay/waku_relay.go [hunk 6]
```diff
@@ -19,6 +20,8 @@ type Topic string
 
 const DefaultWakuTopic Topic = "/waku/2/default-waku/proto"
 
+const WakuRelayProtocolId = libp2pProtocol.ID("/vac/waku/relay/2.0.0-beta2")
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply waku/v2/protocol/relay/waku_relay.go:23]: Thanks, fixed.

File: go.mod [hunk 7]
```diff
@@ -32,3 +32,5 @@ require (
 	google.golang.org/protobuf v1.25.0 // indirect
 	gopkg.in/ini.v1 v1.62.0 // indirect
 )
+
+replace github.com/status-im/go-wakurelay-pubsub => ../go-wakurelay-pubsub
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply go.mod:36]: Done.

File: waku/v2/protocol/store/waku_store.go [hunk 8]
```diff
@@ -215,8 +216,17 @@ func (store *WakuStore) SetMsgProvider(p MessageProvider) {
 	store.msgProvider = p
 }
 
-func (store *WakuStore) Start(h host.Host) {
+func (store *WakuStore) peerListener() {
```
[the remark this answers is not in the document]
Comment (siphiuel) [reply waku/v2/protocol/store/waku_store.go:223]: I've left is as a placeholder for possible future handling. Right now i don't know much about inner workings of PeerStore. Do you want me to remove it?

**Review by richard-ramos — APPROVED** [review]

**Pull request merged**